### ISSUE TITLE
[Bug] DeepSeek-V4-Flash on sm_121 (GB10): deterministic character-level output corruption — quant/build-independent

### ISSUE THREAD

**Issue by dtkennedy23** [ggml-org/llama.cpp#28132]
## Description

Serving a DeepSeek-V4-Flash GGUF on GB10 (sm_121) produces **deterministic, character-level output corruption** when the model reproduces text from its own context. It is 100% reproducible at temperature 0, survives two llama.cpp builds, two different quantizations (MXFP4 and Q2_K_S), and is absent when the same model runs hosted. Output is fluent and plausible — the defect only becomes visible when the model must reproduce exact character sequences (agent workloads: tool calls, paths, shell commands).

Typical failure: asked to echo the exact string

```
ls -la ~/.ssh/config 2>/dev/null | head -20 && hostname && whoami
```

temp 0 returns (deterministic across 5/5 reps):

```
ls -la ~/.ssh/config 2>/dev/nll | head -20 && hostname && whoam
```

The model also *misreads its own prompt*: in reasoning_content it restates the canary dozens of times, each time with different mutations (`2>dev/null`, `2/dev/null`, `2>>/dev/null`, `hosname`, `whoam`, `head 20`), blames the "user's typos", re-reads to resolve the contradiction, produces a new mutation, and spirals until the token budget exhausts (empty content, `finish_reason: length`). This is the generator of agent-facing loop/corruption failures.

## Steps to reproduce

Single request against the server (any of the 4 configs below):

```bash
curl http://<host>:8081/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model": "deepseek-v4-flash",
  "messages": [{"role": "user", "content": "Repeat the following shell command back to me EXACTLY, character for character, on a single line, with nothing before or after it and no code fence:\nls -la ~/.ssh/config 2>/dev/null | head -20 && hostname && whoami"}],
  "temperature": 0,
  "max_tokens": 3000,
  "stream": false
}'
```

Expected: the canary echoed byte-exact. Observed: never byte-exact (0/10 echo + 0/5 tool-call args across every configuration below).

## Exoneration matrix (same 65-char canary, temp 0, n≥5 per arm)

| Variable | Result |
|---|---|
| MTP on / off (`--model-draft`, `--spec-draft-n-max 2`) | corrupts either way |
| `--reasoning-budget 1024` on / off | corrupts either way (budget only moves the spiral from empty-content to content-overflow) |
| temperature 0.0 / 1.0 | corrupts either way; temp 0 is fully deterministic |
| llama.cpp `4df29be4f` (b10454) and `85c55223c` (b10726) | corrupts on both |
| **Quant: MXFP4 GGUF AND Q2_K_S GGUF of the same model** | **corrupts identically on both → the GGUF files and the MXFP4 matmul are not the cause** |
| Samplers | repeat-penalty 1.1 + DRY deployed; failure predates and survives them |
| Tokenizer / template | `/tokenize`→`/detokenize` round-trip byte-exact; `/apply-template` renders the prompt byte-perfect → the corruption is in the forward pass |

**Discriminators:**

| Test | Result |
|---|---|
| Hosted `deepseek-v4-flash` (official DeepSeek API, same prompt) | **10/10 byte-exact**, 86–697 chars reasoning, ~1.1 s/request |
| `gpt-oss-20b` **MXFP4** GGUF, single node, same binary, same sm_121 hardware | **15/15 byte-exact** (echo 10/10, tool args 5/5) → generic MXFP4 MMQ kernels are fine |
| DeepSeek-V4-Flash locally (all configs, 40 requests) | 0 byte-exact |

So: model weights fine (hosted), generic MXFP4 kernels fine (gpt-oss), quant file fine (Q2_K_S also corrupts). What remains specific to this deployment: **the DeepSeek-V4 architecture-specific paths (DSA sparse-attention indexer, MLA, MoE routing) executed on sm_121**, and the 2-node RPC tensor split (could not isolate: loading the full model on one node wedges the GB10 kernel during load — possibly a separate UVM issue).

## Related reports

- MLX #3879: "mxfp4 forward numerically wrong on Blackwell (sm_120/121)… individual ops fine, composed forward diverges ~25%… over 78 layers this compounds into degenerate logits" — same signature on the same silicon in a different runtime.
- llama.cpp #21371: custom CUDA quant kernels producing gibberish on Blackwell (fix: force cuBLAS) — same corruption class.

## Environment

- Hardware: 2× ASUS GX10 (NVIDIA GB10, sm_121, 121 GiB unified each), 200GbE DAC fabric, llama.cpp RPC tensor split 40/60
- OS: Ubuntu 24.04.4 LTS, kernel 6.17.0-1031-nvidia, aarch64, GCC 13.3.0
- Driver: 580.173.02 (CUDA 13.0); built with `-DCMAKE_CUDA_COMPILER=/usr/local/cuda-13.0/bin/nvcc` (`/usr/local/cuda` symlinks to 13.3 — PTX-JIT builds are rejected by the 13.0 driver, so sm_121 SASS-native builds only)
- Build flags: `-DCMAKE_CUDA_ARCHITECTURES=121 -DGGML_CUDA=ON -DGGML_CUDA_FA=ON -DGGML_CUDA_FA_ALL_QUANTS=OFF -DGGML_RPC=ON -DGGML_RPC_RDMA=ON -DBUILD_SHARED_LIBS=ON` (Release)
- Server flags: `-m DeepSeek-V4-Flash-{MXFP4,Q2_K_S}-00001-of-00002.gguf --model-draft mtp-...gguf --spec-draft-n-max 2 --rpc 10.0.0.2:50052 -ngl 99 --tensor-split 40,60 --repeat-penalty 1.1 --dry-multiplier 0.4 --dry-allowed-length 3 --dry-penalty-last-n 256 --reasoning-budget 1024 --alias deepseek-v4-flash --port 8081`

## Ask

Help isolating which sm_121 kernel path (DSA indexer / MLA / MoE for the DeepSeek-V4 architecture) could emit subtly wrong hidden states for context tokens — e.g. a flag to force dequant/cuBLAS/CPU for those paths, or a pointer to known sm_121 numerics issues in those kernels. Full probe harness and logs available on request.

**Comment by dtkennedy23** [comment]
## New discriminator (2026-09-02): RPC transport exonerated — TCP-only rebuild corrupts identically

Rebuilt both nodes at the same commit with `GGML_RPC_RDMA=OFF` (transport verified as `TCP` in the worker log — the production build negotiates RDMA over RoCE), same model, same 40/60 split, same MTP. Full 20-rep probe: **identical corruption** — 0/10 byte-exact echo, 0/5 tool args, same deterministic mutations (`2>/dev/nll`, `whoam`) at temp 0.

So the corruption is NOT in the transport/RPC layer, and it is not the hardware: a vLLM-based community stack on the same sm_121 silicon (NVIDIA forum, Aiden recipe thread) confirmed a byte-exact echo ("the command is 1:1").

That leaves exactly one suspect: the **DeepSeek-V4-Flash arch-specific kernel paths in llama.cpp on sm_121** (DSA sparse-attention indexer / MLA / MoE) producing subtly wrong hidden states deterministically. Happy to run any kernel-level bisect or instrumentation a maintainer suggests.

**Comment by artcashin** [comment]
Data point that corroborates your gpt-oss-20b control and adds a caution about the **stock published CUDA image** on GB10 (sm_121).

**Setup:** single GB10 / DGX Spark (sm_121, driver 580.173.02, CUDA 13.0), `openai/gpt-oss-120b` MXFP4 GGUF (`ggml-org/gpt-oss-120b-GGUF`), `llama-server -ngl 999 --jinja -c 131072`.

**Stock `ghcr.io/ggml-org/llama.cpp:full-cuda` (build b10752, commit b96806d96):** plain-text prompts generate correctly, but any **harmony/channel** prompt degrades into fluent word-salad once generation is non-trivial — deterministic at temp 0 and 0.8, and **byte-identical across two independently-quantized GGUFs (MXFP4 and unsloth Q8_0)**, so not the weights. `/v1/chat/completions` then 500s ("does not match the expected peg-native format") because the PEG parser correctly rejects the garbage. A trivial harmony ask ("Say READY", ~1 real token) stays coherent, which masks it initially.

**Native build fixes it.** Same source commit (b96806d96), built in `nvidia/cuda:13.0.1-devel-ubuntu24.04` with `-DCMAKE_CUDA_ARCHITECTURES=121` (run `--gpus all` so `libcuda.so.1` is present at link time). `cuobjdump` confirms native sm_121 SASS in `libggml-cuda.so`. Same MXFP4 GGUF now emits correct harmony (`analysis`+`final` channels) and clean tool/chat output — 200, not 500.

So on this hardware **gpt-oss (both 20b, per your control, and 120b) is fine on sm_121 with a native build** — the garble was the stock image's codegen falling back wrong on sm_121, not the GB10. That the fix is a *native build* (whereas your DeepSeek-V4-Flash corruption persists **with** a native build) seems to further isolate the DeepSeek case to its arch-specific paths rather than a general sm_121/image problem. It also matches the "compounds over layers" angle from the MLX cross-ref: gpt-oss-20b survived the stock image's flaw where 120b did not.

**Comment by dtkennedy23** [comment]
Thanks — this is a genuinely useful data point, because it closes the loop on the one confound that could have weakened our conclusion.

Our builds are native SASS too (`-DCMAKE_CUDA_ARCHITECTURES=121`, CUDA 13.0 toolchain) — and these boxes give us an independent proof of that: the driver (580.173.02) rejects PTX-JIT entirely ("PTX was compiled with an unsupported toolchain"), so a non-native build won't even load here. Since our corruption persists *with* a native build, your result rules out the stock-image codegen fallback as the explanation for our case — it has to be in the DeepSeek-V4-Flash arch-specific paths themselves (DSA indexer / MLA / MoE) rather than a general sm_121/image codegen problem.

Two small echoes between our setups worth noting: our gpt-oss-20b control was clean on a native build, consistent with your 120b recovering once native; and both your harmony-salad and our echo-corruption are fluent, deterministic-at-temp-0, and quant-independent — the shared signature of subtly wrong hidden states compounding over layers, per the MLX cross-ref.

If you get a chance, running the 65-char echo from the issue body on your native gpt-oss-120b would add one more clean-native control point to the matrix.
